Sources/ContentScopeScripts/dist/contentScope.js

```diff
@@ -3864,78 +3864,230 @@
   });
 
   let adLabelStrings = [];
+  const parser = new DOMParser();
+  const hiddenElements = new WeakMap();
+  const appliedRules = new Set();
 
-  function collapseDomNode (element, type) {
+  /**
+   * Hide DOM element if rule conditions met
+   * @param {HTMLElement} element
+   * @param {Object} rule
+   * @param {HTMLElement} [previousElement]
+   */
+  function collapseDomNode (element, rule, previousElement) {
       if (!element) {
           return
       }
+      const type = rule.type;
+      const alreadyHidden = hiddenElements.has(element);
+
+      if (alreadyHidden) {
+          return
+      }
 
       switch (type) {
       case 'hide':
-          if (!element.hidden) {
-              hideNode(element);
-          }
+          hideNode(element);
           break
       case 'hide-empty':
-          if (!element.hidden && isDomNodeEmpty(element)) {
+          if (isDomNodeEmpty(element)) {
               hideNode(element);
+              appliedRules.add(rule);
           }
           break
       case 'closest-empty':
-          // if element already hidden, continue onto parent element
-          if (element.hidden) {
-              collapseDomNode(element.parentNode, type);
-              break
+          // hide the outermost empty node so that we may unhide if ad loads
+          if (isDomNodeEmpty(element)) {
+              collapseDomNode(element.parentNode, rule, element);
+          } else if (previousElement) {
+              hideNode(previousElement);
+              appliedRules.add(rule);
           }
+          break
+      }
+  }
 
-          if (isDomNodeEmpty(element)) {
-              hideNode(element);
-              collapseDomNode(element.parentNode, type);
+  /**
+   * Unhide previously hidden DOM element if content loaded into it
+   * @param {HTMLElement} element
+   * @param {Object} rule
+   * @param {HTMLElement} [previousElement]
+   */
+  function expandNonEmptyDomNode (element, rule, previousElement) {
+      if (!element) {
+          return
+      }
+      const type = rule.type;
+
+      const alreadyHidden = hiddenElements.has(element);
+
+      switch (type) {
+      case 'hide':
+          // only care about rule types that specifically apply to empty elements
+          break
+      case 'hide-empty':
+          if (alreadyHidden && !isDomNodeEmpty(element)) {
+              unhideNode(element);
+          }
+          break
+      case 'closest-empty':
+          // iterate upwards from matching DOM elements until we arrive at previously
+          // hidden element. Unhide element if it contains visible content.
+          if (alreadyHidden) {
+              if (!isDomNodeEmpty(element)) {
+                  unhideNode(element);
+              }
+              break
+          } else {
+              expandNonEmptyDomNode(element.parentNode, rule);
```

Comment:
These are pretty similar, either make a function or just share the case statement?

```suggestion
      case 'hide-empty':
      case 'closest-empty':
          if (alreadyHidden && !isDomNodeEmpty(element)) {
              unhideNode(element);
          } else if (type === 'closest-empty') {
              // iterate upwards from matching DOM elements until we arrive at previously
              // hidden element. Unhide element if it contains visible content.
              expandNonEmptyDomNode(element.parentNode, rule);
          }
          break
```